FEST-465: isSubsetOf assertion

maciejjaskowski opened the pull request:
Hi Alex,

I have the "isSubsetOf" assertion working :-)

The only thing I couldn't make working is JavaDoc on AbstractIterableAssert.isSubsetOf(). I just couldn't figure out how the {@inheritDoc} magic works.

Apart from that I created tests for both AbstractIterableAssert and Iterables. I named the test methods the best I could. Feel free, however, to make them suite your taste :)

I did also the assumption that we are speaking about sets not multisets. Therefore if we have duplicates on either side, they are seen as one.

Let me know if anything needs my further attention.
Best,
Maciej

maciejjaskowski committed:
FEST-465: fixed

joel-costigliola commented on src/main/java/org/fest/assertions/api/AbstractIterableAssert.java:
```diff
@@ -78,6 +79,12 @@ public final S containsOnly(Object... values) {
     return myself;
   }
 
+
```
Missing javadoc, this is important because it is the api exposed to our users

joel-costigliola commented on src/main/java/org/fest/assertions/internal/Iterables.java:
```diff
@@ -249,6 +251,34 @@ public void assertContainsSequence(AssertionInfo info, Iterable<?> actual, Objec
+  }
+
+  private void checkNotNull(AssertionInfo info, Iterable<?> set) {
+    if (set == null) throw arrayOfValuesToLookForIsNull();
```
the idea is good although I think we should write iterableToLookForIsNull

maciejjaskowski replied:
fixed

joel-costigliola commented on src/test/java/org/fest/assertions/internal/Iterables_assertIsSubsetOf_Test.java:
```diff
@@ -0,0 +1,138 @@
+	  thrown.expectAssertionError(actualIsNull());
+	  iterables.assertIsSubsetOf(someInfo(), null, list());
+  }
+  
```
We should add some tests cases when actaul is not a subset of given value, at least 
should_fail_if_actual_is_not_subset_of_values

maciejjaskowski replied:
fixed

joel-costigliola commented:
Hey Maciej,

Good work ! 
I have only spotted minor missing points.
I can let you fix them or I can do it, it's up to you - let me know.

Cheers,

Joel

maciejjaskowski commented:
I will love to fix them as soon as I can :) (Saturday morning most
probably). Is that ok?

Cheers,
Maciej

On 28 February 2012 21:46, Joel Costigliola
reply@reply.github.com
wrote:

> Hey Maciej,
> 
> Good work !
> I have only spotted minor missing points.
> I can let you fix them or I can do it, whatever you prefer.
> 
> Cheers,
> 
> Joel
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/alexruiz/fest-assert-2.x/pull/2#issuecomment-4225083

joel-costigliola commented:
totally ok ! take your time.

maciejjaskowski commented:
Hi Joel,

I'm not sure what to put as java doc for
AbstractIterableAssert.isSubsetOf or rather -- I have not a clue what
{@inheritDoc} means and therefore I'm not sure _where_ to put a
relevant javadoc. Could you point me to a relevant file ?

Thanks in advance,
Maciej

joel-costigliola commented:
{@inheritDoc} means "reuse javadoc of overridden method".
In our case using {@inheritDoc} does not mean anything since we haven't defined isSubsetOf in a superclass or interface (you can have a look at containsOnly method to see that it reuses the javadoc of ObjectEnumerableAssert interface).

For the time being, there is no javadoc at all on isSubsetOf assertion, you just have to fill javadoc directly AbstractIterableAssert.isSubsetOf and it will be just fine.

This being said, I have a suggestion concerning  the error message. 

Running this test

``` java
Collection<Ring> elvesRings = list(vilya, nenya, narya);
// this one is ok
assertThat(elvesRings).isSubsetOf(ringsOfPower);
// this one results in an error 
assertThat(ringsOfPower).isSubsetOf(elvesRings);
```

causes an with the following message :
Expecting:<[oneRing, vilya, nenya, narya, dwarfRing, manRing]> to be subset of <[vilya, nenya, narya]>

I was thinking it could be helpful to indicate the extra elements that made the assertion to fail, something like
Expecting:<[oneRing, vilya, nenya, narya, dwarfRing, manRing]> to be subset of <[vilya, nenya, narya]> but found those extra elements : [oneRing, dwarfRing, manRing]

What do you think ?

maciejjaskowski commented:
On 4 March 2012 13:53, Joel Costigliola
reply@reply.github.com
wrote:

> {@inheritDoc} means "reuse javadoc of overridden method".
> In our case using {@inheritDoc} does not mean anything since we haven't defined isSubsetOf in a superclass or interface (you can have a look at containsOnly method to see that it reuses the javadoc of ObjectEnumerableAssert interface).
> 
> For the time being, there is no javadoc at all on isSubsetOf assertion, you just have to fill javadoc directly AbstractIterableAssert.isSubsetOf and it will be just fine.

thanks for explanations!

> This being said, I have a suggestion concerning  the error message.
> 
> Running this test
> 
> ``` java
> Collection<Ring> elvesRings = list(vilya, nenya, narya);
> // this one is ok
> assertThat(elvesRings).isSubsetOf(ringsOfPower);
> // this one results in an error
> assertThat(ringsOfPower).isSubsetOf(elvesRings);
> ```
> 
> causes an with the following message :
> Expecting:<[oneRing, vilya, nenya, narya, dwarfRing, manRing]> to be subset of <[vilya, nenya, narya]>
> 
> I was thinking it could be helpful to indicate the extra elements that made the assertion to fail, something like
> Expecting:<[oneRing, vilya, nenya, narya, dwarfRing, manRing]> to be subset of <[vilya, nenya, narya]> but found those extra elements : [oneRing, dwarfRing, manRing]
> 
> What do you think ?

Sounds good to me :) I will fix it with other things as soon as I can.

> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/alexruiz/fest-assert-2.x/pull/2#issuecomment-4309538

maciejjaskowski commented:
I fixed everything you asked me to. But I will not dare to do a pull
request until I have some other 30 minutes to look through my changes
;)

Best,
Maciej

On 4 March 2012 22:49, Maciej Jaśkowski maciej.jaskowski@gmail.com wrote:

> On 4 March 2012 13:53, Joel Costigliola
> reply@reply.github.com
> wrote:
> 
> > {@inheritDoc} means "reuse javadoc of overridden method".
> > In our case using {@inheritDoc} does not mean anything since we haven't defined isSubsetOf in a superclass or interface (you can have a look at containsOnly method to see that it reuses the javadoc of ObjectEnumerableAssert interface).
> > 
> > For the time being, there is no javadoc at all on isSubsetOf assertion, you just have to fill javadoc directly AbstractIterableAssert.isSubsetOf and it will be just fine.
> 
> thanks for explanations!
> 
> > This being said, I have a suggestion concerning  the error message.
> > 
> > Running this test
> > 
> > ``` java
> > Collection<Ring> elvesRings = list(vilya, nenya, narya);
> > // this one is ok
> > assertThat(elvesRings).isSubsetOf(ringsOfPower);
> > // this one results in an error
> > assertThat(ringsOfPower).isSubsetOf(elvesRings);
> > ```
> > 
> > causes an with the following message :
> > Expecting:<[oneRing, vilya, nenya, narya, dwarfRing, manRing]> to be subset of <[vilya, nenya, narya]>
> > 
> > I was thinking it could be helpful to indicate the extra elements that made the assertion to fail, something like
> > Expecting:<[oneRing, vilya, nenya, narya, dwarfRing, manRing]> to be subset of <[vilya, nenya, narya]> but found those extra elements : [oneRing, dwarfRing, manRing]
> > 
> > What do you think ?
> 
> Sounds good to me :) I will fix it with other things as soon as I can.
> 
> > ---
> > 
> > Reply to this email directly or view it on GitHub:
> > https://github.com/alexruiz/fest-assert-2.x/pull/2#issuecomment-4309538

joel-costigliola commented:
Maciej,

As I don't see your modifications, I assume you haven't pushes yet, am I right ?
Don't hesitate to push them so that I can review and integrate them.

Cheers,

Joel

maciejjaskowski committed:
* javadoc added AbstractIterableAssert.isSubsetOf()
* extended error message in ShouldBeSubsetOf (Iterables changed
accordingly)
* a test added to Iterables_assertIsSubsetOf_Test
* Iterables_assertIsSubsetOf_Test refactored a bit (actual is set at the
beginning of each method now. That should improve readability
* readobility of errors thrown from
Iterables.checkIsNotNullAndNotEmpty() improved.

maciejjaskowski committed:
* bugfix
* ShouldBeSubsetOf_create_Test

maciejjaskowski commented:
I did a pull request just now. Then fixed a bug and did a pull request again ;)

To all the fixes you requested I added a slight refactoring of
Iterables_assertIsSubsetOf_Test
and creation of ShouldBeSubsetOf_create_Test

Cheers,
Maciej

On 7 March 2012 21:46, Joel Costigliola
reply@reply.github.com
wrote:

> Maciej,
> 
> As I don't see your modifications, I assume you haven't pushes yet, am I right ?
> Don't hesitate to push them so that I can review and integrate them.
> 
> Cheers,
> 
> Joel
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/alexruiz/fest-assert-2.x/pull/2#issuecomment-4376962

joel-costigliola commented:
Request integrated 

Thanks Maciej !

maciejjaskowski commented:
Great to hear :) Maybe I could undertake some other issues? My spare
time is tight but I'm keen to help :)

Cheers,
Maciej

On 8 March 2012 22:52, Joel Costigliola
reply@reply.github.com
wrote:

> Request integrated
> 
> Thanks Maciej !
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/alexruiz/fest-assert-2.x/pull/2#issuecomment-4403526

joel-costigliola commented:
Hey Maciej

That would be greatly appreciated !

Have a look at and pick one of the opened issues for Fest Assert 2.0 or if
you have other ideas of assertions, create an issue.
In the latter case, please do it in github directly, we are thinking to
move from Jira because of various problems.

Let me know what you wish to work on.

Cheers,

Joel

2012/3/9 Maciej Jaśkowski <
reply@reply.github.com

> Great to hear :) Maybe I could undertake some other issues? My spare
> time is tight but I'm keen to help :)
> 
> Cheers,
> Maciej
> 
> On 8 March 2012 22:52, Joel Costigliola
> reply@reply.github.com
> wrote:
> 
> > Request integrated
> > 
> > Thanks Maciej !
> > 
> > ---
> > 
> > Reply to this email directly or view it on GitHub:
> > https://github.com/alexruiz/fest-assert-2.x/pull/2#issuecomment-4403526
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/alexruiz/fest-assert-2.x/pull/2#issuecomment-4412198
